pkg/virt-handler/cmd-client/client.go

```diff
@@ -112,178 +118,226 @@ func SocketFromUID(baseDir string, uid string) string {
 	return filepath.Join(SocketsDirectory(baseDir), sockFile)
 }
 
-func GetClient(socketPath string) (LauncherClient, error) {
-	conn, err := rpc.Dial("unix", socketPath)
+func NewClient(socketPath string) (LauncherClient, error) {
+	// dial socket
+	conn, err := grpcutil.DialSocket(socketPath)
 	if err != nil {
+		log.Log.Reason(err).Infof("failed to dial cmd socket: %s", socketPath)
 		return nil, err
 	}
 
-	return &VirtLauncherClient{client: conn}, nil
-}
-
-func (c *VirtLauncherClient) Close() {
-	c.client.Close()
+	// create info client and find cmd version to use
+	infoClient := info.NewCmdInfoClient(conn)
+	return NewClientWithInfoClient(infoClient, conn)
 }
 
-func (c *VirtLauncherClient) genericSendCmd(args *Args, cmd string) (*Reply, error) {
-	reply := &Reply{}
+func NewClientWithInfoClient(infoClient info.CmdInfoClient, conn *grpc.ClientConn) (LauncherClient, error) {
+	ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
+	info, err := infoClient.Info(ctx, &info.CmdInfoRequest{})
+	if err != nil {
+		return nil, fmt.Errorf("could not check cmd server version: %v", err)
+	}
+	version, err := com.GetHighestCompatibleVersion(info.SupportedCmdVersions, supportedCmdVersions)
+	if err != nil {
+		return nil, err
+	}
 
-	err := c.client.Call(cmd, args, reply)
-	if IsDisconnected(err) {
-		return reply, err
-	} else if err != nil {
-		msg := fmt.Sprintf("unknown error encountered sending command %s: %s", cmd, err.Error())
-		return reply, fmt.Errorf(msg)
-	} else if reply.Success != true {
-		return reply, fmt.Errorf("server error. command %s failed: %q", cmd, reply.Message)
+	// create cmd client
+	switch version {
+	case 1:
+		client := cmdv1.NewCmdClient(conn)
+		return newV1Client(client, conn), nil
```

Comment:
this is perfect :+1:  great work here, i like the abstraction that's being used. The "LauncherClient" object hide away the fact that multiple cmd-client versions might exist in the future, so virt-handler never has to be aware of this. This is exactly what I was hoping for